### DISCUSSION TITLE
Web RelayDNS Page Not Found Error 404 on http or https connect link

### DISCUSSION THREAD

**Discussion by ExpGy** [Ylianst/MeshCentral#6468] [Issues]
Describe your issue
I enabled the web relayDNS feature, however when I click the http or https links it opens and getting instantly a 404 page does not exist error. Connection path:   ISProuter  --80,443--> TPLink router --80,443-> RasPI(nginx,tactical rmm,meshcentral). Everything is communicating fine(agents), RDP,webRDP working, only http and https not.

Screenshots
![kép](https://github.com/user-attachments/assets/f492d7dd-75e4-48d8-9b18-13d3a9f87003)

Server Software (please complete the following information):
    OS: Debian 12.
    Virtualization: none
    Network: WAN, nginx reverse proxy, letsencrypt, ssl offload
    MC Version: 1.1.32
    Node: 18.20.4

Client Device (please complete the following information):

    Device: PC
    OS: Windows
    Network: Remote to MC, Remote over WAN
    Browser: Edge, Firefox, Google Chrome
    MeshCentralRouter -

Remote Device (please complete the following information):

    Device: PC
    OS: Windows 11
    Network: Local to Meshcentral, Remote over WAN
    Current Core Version (if known): for Ubuntu PC: Aug 29 2022, 2464792450, for Windows PC:Aug 25 2022, 143332014

**config.json:**
{
  "settings": {
    "cert": "mesh.xxxxxxxxx.xx",
    "WANonly": true,
    "minify": 1,
    "port": 4430,
    "aliasPort": 443,
    "redirPort": 800,
    "redirAliasPort": 80,
    "AgentPong": 300,
    "RelayDNS": ["relay1.mesh.xxxxxxxxx.xx",
                 "relay2.mesh.xxxxxxxxx.xx",
                 "relay3.mesh.xxxxxxxxx.xx",
                 "relay4.mesh.xxxxxxxxx.xx",
                 "relay5.mesh.xxxxxxxxx.xx",
                 "relay6.mesh.xxxxxxxxx.xx",
                 "relay7.mesh.xxxxxxxxx.xx",
                 "relay8.mesh.xxxxxxxxx.xx",
                 "relay9.mesh.xxxxxxxxx.xx"],
    "BrowserPong": 300,
    "mstsc": true,
    "allowLoginToken": true,
    "allowFraming": true,
    "agentPing": 35,
    "allowHighQualityDesktop": true,
    "tlsOffload": "100.0.0.250",
    "agentCoreDump": false,
    "compression": true,
    "wsCompression": true,
    "agentWsCompression": true,
    "maxInvalidLogin": {
      "time": 5,
      "count": 5,
      "coolofftime": 30
    },
    "plugins": {
      "enabled": true
    },
    "postgres": {
      "user": "xxxxxxxx",
      "password": "xxxxxxxxxxxxxx",
      "port": "5432",
      "host": "localhost"
    }
  },
  "domains": {
    "": {
      "title": "XXXXXXXXXX",
      "title2": "MESHCENTRAL",
      "newAccounts": false,
      "certUrl": "https://100.0.0.250:443/",
      "geoLocation": true,
      "cookieIpCheck": false,
      "mstsc": true
    }
  }
}

**nginx:**

server {
  listen 80;
  server_name mesh.xxxxxxxxx.xx;
  location / {
     proxy_pass http://100.0.0.250:800/;
     proxy_http_version 1.1;
     proxy_set_header X-Forwarded-Host $host:$server_port;
     proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
     proxy_set_header X-Forwarded-Proto $scheme;
  }
}

server {
    listen 443 ssl;
    server_name mesh.xxxxxxxxx.xx;
    proxy_send_timeout 330s;
    proxy_read_timeout 330s;
    ssl_certificate /etc/letsencrypt/live/xxxxxxxxx.xx/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/xxxxxxxxx.xx/privkey.pem; # managed by Certbot
    ssl_session_cache shared:WEBSSL:10m;
    ssl_ciphers HIGH:!aNULL:!MD5;
    ssl_prefer_server_ciphers on;
    location / {
        proxy_http_version 1.1;
        proxy_pass http://100.0.0.250:4430/;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Host $host:$server_port;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}

**Comment by si458** [comment]
You haven't shared ur config.json file.
How did u setup the relay feature.
Did u use the Port option or the dns option?
The 404 error u are getting is coming from nginx and not meshcentral
So ur nginx isn't setup correctly

**Reply by ExpGy** [reply]
Edited

**Reply by si458** [reply]
Right so u are using the dns option.

So what u need to do is inside ur nginx u need to add all ur relaydns names inside the server name

You could try 
`server_name  mesh.example.org  *.mesh.example.org;`

U also need to add all the relaydns names into ur ssl certificate as well

**Reply by ExpGy** [reply]
"U also need to add all the relaydns names into ur ssl certificate as well"
Already done

"server_name  mesh.example.org  *.mesh.example.org"
Changed

now the server loads my www site not the meshcentral

**Reply by si458** [reply]
I'm not 100% sure as I don't use nginx

I'm only going off the docs i found from nginx?
http://nginx.org/en/docs/http/server_names.html

Have u set the server_name in both 80 and 443?

**Reply by ExpGy** [reply]
Yes both server_name at 80 and 443 are set to *.mesh.mydomain.com

at proxy_pass should i use the * too?
like proxy_pass http://*.mesh.mydomain.com:4430/;

**Reply by ExpGy** [reply]
![kép](https://github.com/user-attachments/assets/96786eb2-8f6f-45fe-95ea-51f8870c4fa9)

**Reply by ExpGy** [reply]
If i remove the wildcard from *.mesh.mydomain.com:4430/; the meshcentral loading again

**Reply by si458** [reply]
u need BOTH the mesh.mydomain.com and *.mesh.mydomain.com in the server_name

**Reply by ExpGy** [reply]
server_name mesh.domain.com *.mesh.domain.com;
both 80 and 443
did not helped, still 404
but now got errors in nginx error log:
2024/10/20 21:57:15 [error] 378249#378249: *770 connect() failed (111: Connection refused) while connecting to upstream, client: 91.82.230.149, server: mesh.fastsupport.hu, request: "GET /agent.ashx
 HTTP/1.1", upstream: "http://127.0.1.1:4430/agent.ashx", host: "mesh.domain.com"

**Reply by ExpGy** [reply]
weird that 127.0.1.1 because its localhost 127.0.0.1

**Reply by si458** [reply]
You should double check ur nginx config and make sure the proxy pass it set to `http://127.0.0.1:4430` if ur running nginx and meshcentral on the same machine 
Also remember to restart nginx between changing its config i believe?

**Reply by ExpGy** [reply]
Changed in nginx proxy_pass http://mesh.domain.com:port/; to http://127.0.0.1:port/; and error 111 gone

**Reply by si458** [reply]
> Changed in nginx proxy_pass http://mesh.domain.com:port/; to http://127.0.0.1:port/; and error 111 gone

So is it working now? I'm confused? 

What does ur nginx config look like now if its not working?

**Reply by ExpGy** [reply]
This is:

```

server {
    if ($host = mesh.domain.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


  listen 80;
  server_name mesh.domain.com *.mesh.domain.com;
  return 301 https://$host$request_uri;

}

server {
    listen 443 ssl;
    server_name  mesh.domain.com *.mesh.domain.com;
    proxy_send_timeout 330s;
    proxy_read_timeout 330s;
    ssl_certificate /etc/letsencrypt/live/*.domain.com/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/*.domain.com/privkey.pem; # managed by Certbot
    ssl_session_cache shared:WEBSSL:10m;
    ssl_ciphers HIGH:!aNULL:!MD5;
    ssl_prefer_server_ciphers on;
    location / {
        proxy_http_version 1.1;
        proxy_pass http://127.0.0.1:4430/;

        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;

        proxy_set_header X-Forwarded-Host $host:$server_port;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }


}

```

And still 404 on http or https link

**Reply by si458** [reply]
Chatgpt suggests, try setting relay1.mesh.domain.com instead of the wildcard to see if its a wildcard issue
Then try accessing the relay1 subdomain and see if it work

**Reply by ExpGy** [reply]
Just found out there is a default.conf where are all the relays listed so:

server {
             root /var/www/html;
             index index.html index.htm index.nginx-debian.html;
             server_name relay2.mesh.fastsupport.hu;
             location / {
                          try_files $uri $uri/ =404;
             }
             listen [::]:443 ssl;
             listen 443 ssl;
             ssl_certificate /etc/letsencrypt/live/*.fastsupport.hu/fullchain.pem;
             ssl_certificate_key /etc/letsencrypt/live/*.fastsupport.hu/privkey.pem;
             include /etc/letsencrypt/options-ssl-nginx.conf;
             ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
}

server {
             if ($host = relay2.mesh.fastsupport.hu) {
                 return 301 https://$host$request_uri;
             }
             listen 80 ;
             listen [::]:80 ;
             server_name relay2.mesh.fastsupport.hu;
             return 404;
}

.....
deleted all the lines fore relays
now getting
![kép](https://github.com/user-attachments/assets/c7b0d7c6-f663-4275-90a0-e3b8d3381ecd)

**Reply by ExpGy** [reply]
It is WORKING!

Had to clear relayX.domain.com related rows form default.conf
And yes the GPT has right! 
so 

server_name mesh.domain.com relay1.mesh.domain.com relay2.mesh.domain.com relay3.mesh.domain.com relay4.mesh.domain.com relay5.mesh.domain.com relay6.mesh.domain.comrelay7.mesh.domain.com relay8.mesh.domain.com relay9.mesh.domain.com;

this way it is working!

I have to mention that I definitely had to prepare a wildcarded certification for the whole thing!

**Thanks for pointing me in the right direction!**

**Reply by ExpGy** [reply]
After cleaning up code wildcard is working too!

```
server {
    if ($host = mesh.domain.com) {
        return 301 https://$host$request_uri;
    }
    listen 80;
    server_name mesh.domain.com *.mesh.domain.com;
    location / {
        proxy_pass http://127.0.0.1:800;
        proxy_http_version 1.1;
        proxy_set_header X-Forwarded-Host $host:$server_port;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}

server {
    listen 443 ssl;
    server_name mesh.domain.com *.mesh.domain.com;
    proxy_send_timeout 330s;
    proxy_read_timeout 330s;
    ssl_certificate /etc/letsencrypt/live/*.domain.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/*.domain.com/privkey.pem;
    ssl_session_cache shared:WEBSSL:10m;
    ssl_ciphers HIGH:!aNULL:!MD5;
    ssl_prefer_server_ciphers on;
    location / {
        proxy_http_version 1.1;
        proxy_pass http://127.0.0.1:4430/;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-Host $host:$server_port;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}


```

**Reply by ExpGy** [reply]
And the working config.json:

```

{
  "settings": {
    "cert": "mesh.domain.com",
    "WANonly": true,
    "minify": 1,
    "port": 4430,
    "aliasPort": 443,
    "redirPort": 800,
    "AgentPong": 300,
    "RelayDNS": ["relay1.mesh.domain.com",
                 "relay2.mesh.domain.com",
                 "relay3.mesh.domain.com",
                 "relay4.mesh.domain.com",
                 "relay5.mesh.domain.com",
                 "relay6.mesh.domain.com",
                 "relay7.mesh.domain.com",
                 "relay8.mesh.domain.com",
                 "relay9.mesh.domain.com"],
    "BrowserPong": 300,
    "mstsc": true,
    "allowLoginToken": true,
    "allowFraming": true,
    "agentPing": 35,
    "allowHighQualityDesktop": true,
    "tlsOffload": "*.mesh.domain.com",
    "agentCoreDump": false,
    "compression": true,
    "wsCompression": true,
    "agentWsCompression": true,
    "maxInvalidLogin": {
      "time": 5,
      "count": 5,
      "coolofftime": 30
    },
    "plugins": {
      "enabled": true
    },
    "postgres": {
      "user": "username",
      "password": "PaSwOrD",
      "port": "5432",
      "host": "localhost"
    }
  },
  "domains": {
    "": {
      "title": "RELAYSUPPORT",
      "title2": "MESHCENTRAL",
      "newAccounts": false,
      "certUrl": "https://mesh.domain.com:443/",
      "geoLocation": true,
      "cookieIpCheck": false,
      "mstsc": true
    }
  }
}

```

**Reply by ExpGy** [reply]
It is also necessary that you have as many DNS entries pointing to your Internet IP address as you want TCP routing at the same time!
Like:
relay1.domain.com  ..... relayX.domain.com

**Reply by si458** [reply]
Glad u got it sorted!

You didn't explain u had a different config file for nginx!

I assume u only have 1 file with everything in

But basically what i explained firstly was correct, u just had a hidden config i wasn't aware of, haha

Yes the relay will basically round Robin the dns entries or you can use a single dns,
Sadly meshcentral isn't setup yet for a wildcard dns for the relaydns but the is an open issue asking for it tho

Also yes u should see a Not Found 404 message when it talks to meshcentral, but u shouldn't see the nginx text at the bottom
